examples/flask/app.py

```diff
@@ -0,0 +1,116 @@
+from flask import (
+    Flask,
+    send_from_directory,
+    request,
+    make_response,
+    redirect,
+    abort,
+)
+import os
+from fetch_static import fetch_static
+from sgid_client import SgidClient, generate_pkce_pair
+from dotenv import load_dotenv
+from uuid import uuid4
+from urllib.parse import urlencode, parse_qs
+import webbrowser
+
+load_dotenv()
+
+fetch_static.fetch_static_files()
+
+# In-memory store for user session data
+# In a real application, this would be a database.
+session_data = {}
+SESSION_COOKIE_NAME = "exampleAppSession"
+
+app = Flask(__name__)
+sgid_client = SgidClient(
+    client_id=os.getenv("SGID_CLIENT_ID"),
+    client_secret=os.getenv("SGID_CLIENT_SECRET"),
+    private_key=os.getenv("SGID_PRIVATE_KEY"),
+    redirect_uri="http://localhost:2000/api/callback",
+)
+
+
+@app.route("/api/auth-url")
+def get_auth_url():
+    ice_cream_selection = request.args.get("icecream")
+    session_id = str(uuid4())
+    # Use search params to store state so other key-value pairs
+    # can be added easily
+    state = urlencode(
+        {
+            "icecream": ice_cream_selection,
+        }
+    )
+    # We pass the user's ice cream preference as the state,
+    # so after they log in, we can display it together with the
+    # other user info.
+    pkce_pair = generate_pkce_pair()
+    auth_url_and_nonce = sgid_client.authorization_url(
+        state=state, code_challenge=pkce_pair["code_challenge"]
+    )
+    session_data[session_id] = {
+        "state": state,
+        "nonce": auth_url_and_nonce["nonce"],
+        "code_verifier": pkce_pair["code_verifier"],
+    }
+    res = make_response({"url": auth_url_and_nonce["url"]})
```

Comment:
This happens because of fetch. If we are using 302 method, the front-end will navigate to `domain/api/auth-url` and then cookies will be set to the domain and 302 happens. If you are using fetch while be responded 302, then yes, CORS will occur.

Reply:
haha I didn't know about the CORS issue. previously my thinking was that returning the URL offers the client slightly more flexibility for their login implementation. e.g. if they want their login button to be more responsive, they can prefetch the auth URL in the background when the page loads, so clicking the login button can bring the user directly to the sgID QR code without requiring an additional API call to fetch the URL.

but since the CORS issue is a thing, I think we should probably stick to current implementation?

Reply:
See step 3 and 6. https://docs.id.gov.sg/integrations-with-sgid/typescript-javascript/next.js-client-side-rendering. 

The difference between 200 and 302 are:
200 - allows different origin. Navigation handles by SPA. 
302 - enforces FE and BE same origin. Navigation handles by browser. (CORS occurs if and only if SPA is trying to handle the navigation which shouldn't)

Prefetch works for both but since state is a queryParam, @raynerljm has disabled the prefetch. 

Happy to standardise on one implementation.

Reply:
> This happens because of fetch. If we are using 302 method, the front-end will have to navigate to `domain/api/auth-url` and then cookies will be set to the domain and 302 happens. If you are using fetch while being responded with 302, then yes, CORS will occur.

Yeah, this is what I was trying to describe. The in the SPA frontend example, we need to use fetch with `credentials: true` for the cookie to be sent. This means that on the backend, we need to configure our CORS middleware with both `credentials: true` plus set the frontend origin to be allowed. Not sure if there is an alternative for our backend for frontend (BFF) examples, which mean that we need to standardize Next.js CSR example to be the same as our BFF examples

Reply:
This happens if and only if FE and BE are on different origin. If the plan is to have FE and BE on different origin, then yes, 200 response with auth url is the only solution. But if our plan is to have FE and BE to be on same origin, then we have two options.

Reply:
For BFF example, really depends if we want same origin or different.

Reply:
i'm actually not very clear how the CORS error crops up + why 302 enforces same FE BE origin - I'll find one of you to understand this discussion better haha

Reply:
okie so I'm just gonna standardise this to what https://github.com/opengovsg/sgid-client/pull/50 is doing, ie return a 200 with the URL. not worth overthinking this imo